File: src/main/java/akka/first/app/mapreduce/actors/ReduceActor.java
package akka.first.app.mapreduce.actors;

import akka.actor.ActorRef;
import akka.actor.UntypedActor;
import akka.first.app.mapreduce.messages.MapData;
import akka.first.app.mapreduce.messages.ReduceData;
import akka.first.app.mapreduce.messages.WordCount;

import java.util.HashMap;
import java.util.List;

/**
 * Created by 98379720172 on 06/05/2016.
 */
public class ReduceActor extends UntypedActor{
    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof MapData) {
            MapData mapData = (MapData) message;
            // reduce the incoming data and forward the result to Master actor
            getSender().tell(reduce(mapData.getDataList()), ActorRef.noSender());
        } else
            unhandled(message);
    }

    private ReduceData reduce(List<WordCount> dataList) {
        HashMap<String, Integer> reducedMap = new HashMap<String,Integer>();

        for (WordCount wordCount : dataList) {
            if (reducedMap.containsKey(wordCount.getWord())) {
                Integer value = (Integer)reducedMap.get(wordCount.getWord());
                value++;
                reducedMap.put(wordCount.getWord(), value);
            } else {
                reducedMap.put(wordCount.getWord(),Integer.valueOf(1));
            }
        }

        return new ReduceData(reducedMap);
    }
}

File: src/main/scala-2.11/scala/akka/first/app/mapreduce/actors/ReduceActor.scala
package scala.akka.first.app.mapreduce.actors

/**
  * Created by dsilv on 07/05/2016.
  */
import akka.actor.Actor

import scala.akka.first.app.mapreduce.{MapData, ReduceData, WordCount}
import scala.collection.immutable.Map


class ReduceActor extends Actor {

  def receive: Receive = {
    case MapData(dataList) =>
      sender ! reduce(dataList)
  }

  def reduce(words: IndexedSeq[WordCount]): ReduceData = ReduceData {
    words.foldLeft(Map.empty[String, Int]) { (index, words) =>
      if (index contains words.word)
        index + (words.word -> (index.get(words.word).get + 1))
      else
        index + (words.word -> 1)
    }
  }
}
